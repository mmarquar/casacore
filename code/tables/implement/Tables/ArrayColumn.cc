//# ArrayColumn.cc: Access to an array table column with arbitrary data type
//# Copyright (C) 1994,1995,1996,1997
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <aips/Tables/ArrayColumn.h>
#include <aips/Tables/Table.h>
#include <aips/Arrays/Array.h>
#include <aips/Arrays/ArrayIter.h>
#include <aips/Lattices/IPosition.h>
#include <aips/Lattices/Slicer.h>
#include <aips/Utilities/String.h>
#include <aips/Utilities/ValTypeId.h>
#include <aips/Tables/TableError.h>


template<class T>
ROArrayColumn<T>::ROArrayColumn()
: ROTableColumn (),
  reaskAccessSlice_p       (new Bool(True)),
  reaskAccessColumn_p      (new Bool(True)),
  reaskAccessColumnSlice_p (new Bool(True)),
  canAccessSlice_p         (new Bool(False)),
  canAccessColumn_p        (new Bool(False)),
  canAccessColumnSlice_p   (new Bool(False)),
  canChangeShape_p         (False)
{}

template<class T>
ROArrayColumn<T>::ROArrayColumn (const Table& tab,
				 const String& columnName)
: ROTableColumn (tab, columnName),
  reaskAccessSlice_p       (new Bool(True)),
  reaskAccessColumn_p      (new Bool(True)),
  reaskAccessColumnSlice_p (new Bool(True)),
  canAccessSlice_p         (new Bool(False)),
  canAccessColumn_p        (new Bool(False)),
  canAccessColumnSlice_p   (new Bool(False))
{
    checkDataType();
    canChangeShape_p = baseColPtr_p->canChangeShape();
}

template<class T>
ROArrayColumn<T>::ROArrayColumn (const ROTableColumn& column)
: ROTableColumn (column),
  reaskAccessSlice_p       (new Bool(True)),
  reaskAccessColumn_p      (new Bool(True)),
  reaskAccessColumnSlice_p (new Bool(True)),
  canAccessSlice_p         (new Bool(False)),
  canAccessColumn_p        (new Bool(False)),
  canAccessColumnSlice_p   (new Bool(False))
{
    checkDataType();
    canChangeShape_p = baseColPtr_p->canChangeShape();
}

template<class T>
ROArrayColumn<T>::ROArrayColumn (const ROArrayColumn<T>& that)
: ROTableColumn (that),
  reaskAccessSlice_p       (new Bool (*that.reaskAccessSlice_p)),
  reaskAccessColumn_p      (new Bool (*that.reaskAccessColumn_p)),
  reaskAccessColumnSlice_p (new Bool (*that.reaskAccessColumnSlice_p)),
  canAccessSlice_p         (new Bool (*that.canAccessSlice_p)),
  canAccessColumn_p        (new Bool (*that.canAccessColumn_p)),
  canAccessColumnSlice_p   (new Bool (*that.canAccessColumnSlice_p)),
  canChangeShape_p         (that.canChangeShape_p)
{}

template<class T>
ROTableColumn* ROArrayColumn<T>::clone() const
{
    return new ROArrayColumn<T> (*this);
}

template<class T>
void ROArrayColumn<T>::reference (const ROArrayColumn<T>& that)
{
    ROTableColumn::reference (that);
    *reaskAccessSlice_p       = *that.reaskAccessSlice_p;
    *reaskAccessColumn_p      = *that.reaskAccessColumn_p;
    *reaskAccessColumnSlice_p = *that.reaskAccessColumnSlice_p;
    *canAccessSlice_p         = *that.canAccessSlice_p;
    *canAccessColumn_p        = *that.canAccessColumn_p;
    *canAccessColumnSlice_p   = *that.canAccessColumnSlice_p;
    canChangeShape_p          = that.canChangeShape_p;
}

template<class T>
ROArrayColumn<T>::~ROArrayColumn()
{
    delete reaskAccessSlice_p;
    delete reaskAccessColumn_p;
    delete reaskAccessColumnSlice_p;
    delete canAccessSlice_p;
    delete canAccessColumn_p;
    delete canAccessColumnSlice_p;
}


template<class T>
void ROArrayColumn<T>::checkDataType() const
{
    //# Check if the data type matches.
    const ColumnDesc& cd = baseColPtr_p->columnDesc();
    DataType dtype = cd.dataType();
    if (dtype != ValType::getType((T*)0)  ||  !cd.isArray()) {
	throw (TableInvDT (" in ROArrayColumn ctor for column " + cd.name()));
    }
    if (dtype == TpOther) {
	if (cd.dataTypeId() != valDataTypeId((T*)0)) {
	    throw (TableInvDT (" in ROArrayColumn ctor for column "
			       + cd.name() + "; using data type id "
			       + valDataTypeId((T*)0)
			       + ", expected " + cd.dataTypeId()));
	}
    }
}


template<class T>
Array<T> ROArrayColumn<T>::operator() (uInt rownr) const
{
    Array<T> arr;
    get (rownr, arr);
    return arr;
}


template<class T>
void ROArrayColumn<T>::get (uInt rownr, Array<T>& arr, Bool resize) const
{
    TABLECOLUMNCHECKROW(rownr); 
    //# Check the array conformance.
    //# Resize the array if empty or if resize flag is set.
    IPosition shp = shape(rownr);
    if (! shp.isEqual (arr.shape())) {
	if (resize  ||  arr.nelements() == 0) {
	    arr.resize (shp);
	}else{
	    throw (TableArrayConformanceError("ArrayColumn::get"));
	}
    }
    baseColPtr_p->get (rownr, &arr);
}

template<class T>
Array<T> ROArrayColumn<T>::getSlice (uInt rownr, const Slicer& ns) const
{
    Array<T> arr;
    getSlice (rownr, ns, arr);
    return arr;
}

template<class T>
void ROArrayColumn<T>::getSlice (uInt rownr, const Slicer& ns,
				 Array<T>& arr, Bool resize) const
{
    TABLECOLUMNCHECKROW(rownr);
    //# Check the array conformance.
    //# Extend the array if empty.
    IPosition arrayShape (shape(rownr));
    IPosition blc,trc,inc;
    IPosition shp = ns.inferShapeFromSource (arrayShape, blc,trc,inc);
    if (! shp.isEqual (arr.shape())) {
	if (resize  ||  arr.nelements() == 0) {
	    arr.resize (shp);
	}else{
	    throw (TableArrayConformanceError("ArrayColumn::getSlice"));
	}
    }
    //# Ask if we can access the slice (if that is not known yet).
    if (*reaskAccessSlice_p) {
	*canAccessSlice_p = baseColPtr_p->canAccessSlice (*reaskAccessSlice_p);
    }
    //# Access the slice if possible.
    //# Otherwise get the entire array and slice it.
    if (*canAccessSlice_p) {
	baseColPtr_p->getSlice (rownr, ns, &arr);
    }else{
	Array<T> array(arrayShape);
	baseColPtr_p->get (rownr, &array);
	arr = array(blc, trc, inc);
    }
}

template<class T>
Array<T> ROArrayColumn<T>::getColumn() const
{
    Array<T> arr;
    getColumn (arr);
    return arr;
}

template<class T>
void ROArrayColumn<T>::getColumn (Array<T>& arr, Bool resize) const
{
    uInt nrrow = nrow();
    //# Take shape of array in first row.
    IPosition shp;
    if (nrrow > 0) {
	shp = shape(0);
    }
    //# Total shape is array shape plus nr of table rows.
    shp.resize (shp.nelements() + 1);
    shp(shp.nelements()-1) = nrrow;
    if (! shp.isEqual (arr.shape())) {
	if (resize  ||  arr.nelements() == 0) {
	    arr.resize (shp);
	}else{
	    throw (TableArrayConformanceError("ArrayColumn::getColumn"));
	}
    }
    //# Ask if we can access the column (if that is not known yet).
    if (*reaskAccessColumn_p) {
	*canAccessColumn_p = baseColPtr_p->canAccessArrayColumn
	                                           (*reaskAccessColumn_p);
    }
    //# Access the column if possible.
    //# Otherwise fill the entire array by looping through all cells.
    if (*canAccessColumn_p) {
	baseColPtr_p->getArrayColumn (&arr);
    }else{
	ArrayIterator<T> iter(arr, arr.ndim()-1);
	for (uInt rownr=0; rownr<nrrow; rownr++) {
	    baseColPtr_p->get (rownr, &(iter.array()));
	    iter.next();
	}
    }
}

template<class T>
Array<T> ROArrayColumn<T>::getColumn (const Slicer& ns) const
{
    Array<T> arr;
    getColumn (ns, arr);
    return arr;
}

template<class T>
void ROArrayColumn<T>::getColumn (const Slicer& ns,
				  Array<T>& arr, Bool resize) const
{
    uInt nrrow = nrow();
    //# Use shape of array in first row.
    IPosition shp, blc,trc,inc;
    if (nrrow > 0) {
	shp = ns.inferShapeFromSource (shape(0), blc,trc,inc);
    }
    //# Total shape is slice shape plus nr of table rows.
    shp.resize (shp.nelements() + 1);
    shp(shp.nelements()-1) = nrrow;
    if (! shp.isEqual (arr.shape())) {
	if (resize  ||  arr.nelements() == 0) {
	    arr.resize (shp);
	}else{
	    throw (TableArrayConformanceError("ArrayColumn::getColumn"));
	}
    }
    //# Ask if we can access the column slice (if that is not known yet).
    if (*reaskAccessColumnSlice_p) {
	*canAccessColumnSlice_p = baseColPtr_p->canAccessColumnSlice
	                                       (*reaskAccessColumnSlice_p);
    }
    //# Access the column slice if possible.
    //# Otherwise fill the entire array by looping through all cells.
    if (*canAccessColumnSlice_p) {
	baseColPtr_p->getColumnSlice (ns, &arr);
    }else{
	ArrayIterator<T> iter(arr, arr.ndim()-1);
	for (uInt rownr=0; rownr<nrrow; rownr++) {
	    getSlice (rownr, ns, iter.array());
	    iter.next();
	}
    }
}


template<class T>
Array<T> ROArrayColumn<T>::getColumnRange (const Slicer& rowRange) const
{
    Array<T> arr;
    getColumnRange (rowRange, arr);
    return arr;
}

template<class T>
void ROArrayColumn<T>::getColumnRange (const Slicer& rowRange,
				       Array<T>& arr, Bool resize) const
{
    uInt nrrow = nrow();
    IPosition shp, blc, trc, inc;
    shp = rowRange.inferShapeFromSource (IPosition(1,nrrow), blc, trc, inc);
    //# When the entire column is accessed, use that function.
    if (shp(0) == nrrow) {
	getColumn (arr, resize);
	return;
    }
    //# Take shape of array in first row.
    nrrow = shp(0);
    IPosition arrshp;
    if (nrrow > 0) {
	arrshp = shape(blc(0));
    }
    //# Total shape is array shape plus nr of table rows.
    arrshp.resize (arrshp.nelements() + 1);
    arrshp(arrshp.nelements()-1) = nrrow;
    if (! arrshp.isEqual (arr.shape())) {
	if (resize  ||  arr.nelements() == 0) {
	    arr.resize (arrshp);
	}else{
	    throw (TableArrayConformanceError("ArrayColumn::getColumnRange"));
	}
    }
    //# Fill the entire array by looping through all cells.
    uInt rownr = blc(0);
    uInt incr = inc(0);
    ArrayIterator<T> iter(arr, arr.ndim()-1);
    for (uInt i=0; i<nrrow; i++) {
	baseColPtr_p->get (rownr, &(iter.array()));
	rownr += incr;
	iter.next();
    }
}




template<class T>
ArrayColumn<T>::ArrayColumn()
: ROTableColumn    (),
  ROArrayColumn<T> (),
  TableColumn      ()
{}

template<class T>
ArrayColumn<T>::ArrayColumn (const Table& tab, const String& columnName)
: ROTableColumn    (tab, columnName),
  ROArrayColumn<T> (tab, columnName),
  TableColumn      (tab, columnName)
{}

template<class T>
ArrayColumn<T>::ArrayColumn (const TableColumn& column)
: ROTableColumn    (column),
  ROArrayColumn<T> (column),
  TableColumn      (column)
{}

template<class T>
ArrayColumn<T>::ArrayColumn (const ArrayColumn<T>& that)
: ROTableColumn    (that),
  ROArrayColumn<T> (that),
  TableColumn      (that)
{}

template<class T>
ROTableColumn* ArrayColumn<T>::clone() const
{
    return new ArrayColumn<T> (*this);
}

template<class T>
void ArrayColumn<T>::reference (const ArrayColumn<T>& that)
    { ROArrayColumn<T>::reference (that); }

template<class T>
ArrayColumn<T>::~ArrayColumn()
{}


template<class T>
void ArrayColumn<T>::setShape (uInt rownr, const IPosition& shape)
{
    TABLECOLUMNCHECKROW(rownr); 
    //# Set shape if not defined yet or if changed (if possible).
    //# Throw exception if already defined with a different shape.
    if (canChangeShape_p  ||  !isDefined(rownr)) {
	baseColPtr_p->setShape (rownr, shape);
    }else{
	if (! shape.isEqual (baseColPtr_p->shape (rownr))) {
	    throw (TableInvOper
		   ("ArrayColumn::setShape; shape cannot be changed"));
	}
    }
}
	
template<class T>
void ArrayColumn<T>::setShape (uInt rownr, const IPosition& shape,
			       const IPosition& tileShape)
{
    TABLECOLUMNCHECKROW(rownr); 
    //# Only set shape if not defined yet.
    //# Throw exception if already defined with a different shape.
    if (canChangeShape_p  ||  !isDefined(rownr)) {
	baseColPtr_p->setShape (rownr, shape, tileShape);
    }else{
	if (! shape.isEqual (baseColPtr_p->shape (rownr))) {
	    throw (TableInvOper
		   ("ArrayColumn::setShape; shape cannot be changed"));
	}
    }
}
	
template<class T>
void ArrayColumn<T>::put (uInt rownr, const Array<T>& arr)
{
    TABLECOLUMNCHECKROW(rownr); 
    //# Define the shape if not defined yet.
    //# If defined, check if shape conforms.
    if (!isDefined(rownr)) {
	setShape (rownr, arr.shape());
    }else{
	if (! arr.shape().isEqual (baseColPtr_p->shape (rownr))) {
	    if (!canChangeShape_p) {
		throw (TableArrayConformanceError("ArrayColumn::put"));
	    }
	    setShape (rownr, arr.shape());
	}
    }
    baseColPtr_p->put (rownr, &arr);
}

template<class T>
void ArrayColumn<T>::putSlice (uInt rownr, const Slicer& ns,
			       const Array<T>& arr)
{
    TABLECOLUMNCHECKROW(rownr); 
    //# Check the array conformance.
    IPosition arrayShape (shape(rownr));
    IPosition blc,trc,inc;
    IPosition shp = ns.inferShapeFromSource (arrayShape, blc,trc,inc);
    if (! shp.isEqual (arr.shape())) {
	throw (TableArrayConformanceError("ArrayColumn::putSlice"));
    }
    //# Ask if we can access the slice (if that is not known yet).
    if (*reaskAccessSlice_p) {
	*canAccessSlice_p = baseColPtr_p->canAccessSlice (*reaskAccessSlice_p);
    }
    //# Access the slice if possible.
    //# Otherwise get the entire array, put the slice and put it back.
    if (*canAccessSlice_p) {
	baseColPtr_p->putSlice (rownr, ns, &arr);
    }else{
	Array<T> array(arrayShape);
	baseColPtr_p->get (rownr, &array);
	array(blc, trc, inc) = arr;
	baseColPtr_p->put (rownr, &array);
    }
}

template<class T>
void ArrayColumn<T>::put (uInt thisRownr, const ROTableColumn& that,
			  uInt thatRownr)
{
    TableColumn::put (thisRownr, that, thatRownr);
}

template<class T>
void ArrayColumn<T>::putColumn (const Array<T>& arr)
{
    //# First check if number of rows matches.
    uInt nrrow = nrow();
    IPosition shp  = arr.shape();
    uInt last = shp.nelements() - 1;
    if (shp(last) != nrrow) {
	throw (TableArrayConformanceError("ArrayColumn::putColumn (nrrow)"));
    }
    //# Remove #rows from shape to get the shape of each cell.
    shp.resize (last);
    //# When the array is fixed shape, check if the shape matches.
    if ((columnDesc().options() & ColumnDesc::FixedShape)
	                                     == ColumnDesc::FixedShape) {
	if (! shp.isEqual (shapeColumn())) {
	    throw (TableArrayConformanceError("ArrayColumn::putColumn"));
	}
    }else{
	//# Otherwise set the shape of each cell (as far as needed).
	for (uInt i=0; i<nrrow; i++) {
	    setShape (i, shp);
	}
    }
    //# Ask if we can access the column (if that is not known yet).
    if (*reaskAccessColumn_p) {
	*canAccessColumn_p = baseColPtr_p->canAccessArrayColumn
	                                             (*reaskAccessColumn_p);
    }
    //# Access the column if possible.
    //# Otherwise put the entire array by looping through all cells.
    if (*canAccessColumn_p) {
	baseColPtr_p->putArrayColumn (&arr);
    }else{
	ReadOnlyArrayIterator<T> iter(arr, arr.ndim()-1);
	for (uInt rownr=0; rownr<nrrow; rownr++) {
	    baseColPtr_p->put (rownr, &(iter.array()));
	    iter.next();
	}
    }
}

template<class T>
void ArrayColumn<T>::putColumn (const Slicer& ns, const Array<T>& arr)
{
    uInt nrrow = nrow();
    //# When the array is fixed shape, check if the shape matches.
    if ((columnDesc().options() & ColumnDesc::FixedShape)
	                                     == ColumnDesc::FixedShape) {
	IPosition blc,trc,inc;
	IPosition shp = ns.inferShapeFromSource (shapeColumn(), blc,trc,inc);
	//# Total shape is slice shape plus nr of table rows.
	shp.resize (shp.nelements() + 1);
	shp(shp.nelements()-1) = nrrow;
	if (! shp.isEqual(arr.shape())) {
	    throw (TableArrayConformanceError("ArrayColumn::putColumn"));
	}
    }
    //# Ask if we can access the column slice (if that is not known yet).
    if (*reaskAccessColumnSlice_p) {
	*canAccessColumnSlice_p = baseColPtr_p->canAccessColumnSlice
	                                       (*reaskAccessColumnSlice_p);
    }
    //# Access the column slice if possible.
    //# Otherwise put the entire array by looping through all cells.
    if (*canAccessColumnSlice_p) {
	baseColPtr_p->putColumnSlice (ns, &arr);
    }else{
	ReadOnlyArrayIterator<T> iter(arr, arr.ndim()-1);
	for (uInt rownr=0; rownr<nrrow; rownr++) {
	    putSlice (rownr, ns, iter.array());
	    iter.next();
	}
    }
}


template<class T>
void ArrayColumn<T>::putColumnRange (const Slicer& rowRange,
				     const Array<T>& arr)
{
    uInt nrrow = nrow();
    IPosition shp, blc, trc, inc;
    shp = rowRange.inferShapeFromSource (IPosition(1,nrrow), blc, trc, inc);
    //# When the entire column is accessed, use that function.
    if (shp(0) == nrrow) {
	putColumn (arr);
	return;
    }
    //# First check if number of rows matches.
    nrrow = shp(0);
    IPosition arrshp  = arr.shape();
    uInt last = arrshp.nelements() - 1;
    if (arrshp(last) != nrrow) {
	throw (TableArrayConformanceError(
	                              "ArrayColumn::putColumnRange (nrrow)"));
    }
    //# Remove #rows from shape to get the shape of each cell.
    arrshp.resize (last);
    //# When the array is fixed shape, check if the shape matches.
    if ((columnDesc().options() & ColumnDesc::FixedShape)
	                                     == ColumnDesc::FixedShape) {
	if (! arrshp.isEqual (shapeColumn())) {
	    throw (TableArrayConformanceError("ArrayColumn::putColumn"));
	}
    }else{
	//# Otherwise set the shape of each cell (as far as needed).
	uInt rownr = blc(0);
	uInt incr = inc(0);
	for (uInt i=0; i<nrrow; i++) {
	    setShape (rownr, arrshp);
	    rownr += incr;
	}
    }
    //# Put the entire array by looping through all cells.
    uInt rownr = blc(0);
    uInt incr = inc(0);
    ReadOnlyArrayIterator<T> iter(arr, arr.ndim()-1);
    for (uInt i=0; i<nrrow; i++) {
	baseColPtr_p->put (rownr, &(iter.array()));
	rownr += incr;
	iter.next();
    }
}


template<class T>
void ArrayColumn<T>::put (uInt thisRownr, const ROArrayColumn<T>& that,
			  uInt thatRownr)
{
    put (thisRownr, that(thatRownr));
}

//# This is a very simple implementation.
//# However, it does not need to be more fancy, since an array operation
//# is already much more expensive than the virtual function calls
//# involved in each loop iteration.
template<class T>
void ArrayColumn<T>::fillColumn (const Array<T>& value)
{
    uInt nrrow = nrow();
    for (uInt i=0; i<nrrow; i++) {
	put (i, value);
    }
}

template<class T>
void ArrayColumn<T>::putColumn (const ROArrayColumn<T>& that)
{
    //# Check the column lengths.
    uInt nrrow = nrow();
    if (nrrow != that.nrow()) {
	throw (TableConformanceError ("ArrayColumn<T>::putColumn"));
    }
    for (uInt i=0; i<nrrow; i++) {
	put (i, that, i);
    }
}
