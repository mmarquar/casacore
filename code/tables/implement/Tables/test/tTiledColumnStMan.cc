//# tTiledColumnStMan.cc: Test program for the TiledColumnStMan classes
//# Copyright (C) 1994,1995,1996
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This program is free software; you can redistribute it and/or modify it
//# under the terms of the GNU General Public License as published by the Free
//# Software Foundation; either version 2 of the License, or (at your option)
//# any later version.
//#
//# This program is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
//# more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with this program; if not, write to the Free Software Foundation, Inc.,
//# 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <aips/Tables/TableDesc.h>
#include <aips/Tables/SetupNewTab.h>
#include <aips/Tables/Table.h>
#include <aips/Tables/ScaColDesc.h>
#include <aips/Tables/ArrColDesc.h>
#include <aips/Tables/ScalarColumn.h>
#include <aips/Tables/ArrayColumn.h>
#include <aips/Tables/TiledColumnStMan.h>
#include <aips/Tables/TiledStManAccessor.h>
#include <aips/Arrays/Vector.h>
#include <aips/Arrays/Matrix.h>
#include <aips/Arrays/Cube.h>
#include <aips/Arrays/ArrayMath.h>
#include <aips/Arrays/ArrayLogical.h>
#include <aips/Arrays/ArrayUtil.h>
#include <aips/Arrays/ArrayIter.h>
#include <aips/Lattices/Slicer.h>
#include <aips/Exceptions/Error.h>
#include <iostream.h>

// <summary>
// Test program for the TiledColumnStMan class.
// </summary>

// This program tests the class TiledColumnStMan and related classes.
// The results are written to stdout. The script executing this program,
// compares the results with the reference output file.

void writeFixed();
void readTable();

main () {
    try {
	writeFixed();
	readTable();
    } catch (AipsError x) {
	cout << "Caught an exception: " << x.getMesg() << endl;
	return 1;
    } end_try;
    return 0;                           // exit with success status
}

// First build a description.
void writeFixed()
{
    // Build the table description.
    TableDesc td ("", "1", TableDesc::Scratch);
    td.addColumn (ArrayColumnDesc<float>  ("Pol", IPosition(1,16),
					   ColumnDesc::FixedShape));
    td.addColumn (ArrayColumnDesc<float>  ("Freq", 1, ColumnDesc::FixedShape));
    td.addColumn (ScalarColumnDesc<float> ("Time"));
    td.addColumn (ArrayColumnDesc<float>  ("Data", 2, ColumnDesc::FixedShape));
    td.addColumn (ArrayColumnDesc<float>  ("Weight", IPosition(2,16,20),
					   ColumnDesc::FixedShape));
    td.defineHypercolumn ("TSMExample",
			  3,
			  stringToVector ("Data,Weight"),
			  stringToVector ("Pol,Freq,Time"));
    
    // Now create a new table from the description.
    SetupNewTable newtab("tTiledColumnStMan_tmp.data", td, Table::New);
    // Create a storage manager for it.
    TiledColumnStMan sm1 ("TSMExample", IPosition(2,5,6));
    newtab.setShapeColumn ("Freq", IPosition(1,20));
    newtab.setShapeColumn ("Data", IPosition(2,16,20));
    newtab.bindAll (sm1);
    Table table(newtab);

    Vector<float> freqValues(20);
    Vector<float> polValues(16);
    indgen (freqValues.ac(), float(200));
    indgen (polValues.ac(), float(300));
    float timeValue;
    timeValue = 34;
    ArrayColumn<float> freq (table, "Freq");
    ArrayColumn<float> pol (table, "Pol");
    ArrayColumn<float> data (table, "Data");
    ArrayColumn<float> weight (table, "Weight");
    ScalarColumn<float> time (table, "Time");
    Matrix<float> array(IPosition(2,16,20));
    Matrix<float> result(IPosition(2,16,20));
    uInt i;
    indgen (array.ac());
    for (i=0; i<51; i++) {
	table.addRow();
	data.put (i, array);
	weight.put (i, array.ac()+float(100));
	time.put (i, timeValue);
	array.ac() += float(200);
	timeValue += 5;
    }
    freq.put (0, freqValues);
    pol.put (0, polValues);
    timeValue = 34;
    indgen (array.ac());
    for (i=0; i<table.nrow(); i++) {
	data.get (i, result);
	if (! allEQ (array.ac(), result.ac())) {
	    cout << "mismatch in data row " << i << endl;
	}
	weight.get (i, result);
	if (! allEQ (array.ac() + float(100), result.ac())) {
	    cout << "mismatch in weight row " << i << endl;
	}
	if (! allEQ (freq(i), freqValues.ac())) {
	    cout << "mismatch in freq row " << i << endl;
	}
	if (! allEQ (pol(i), polValues.ac())) {
	    cout << "mismatch in pol row " << i << endl;
	}
	if (time(i) != timeValue) {
	    cout << "mismatch in time row " << i << endl;
	}
	array.ac() += float(200);
	timeValue += 5;
    }
    ROTiledStManAccessor accessor (table, "TSMExample");
    accessor.showCacheStatistics (cout);
}

void readTable()
{
    Table table("tTiledColumnStMan_tmp.data");
    ROTiledStManAccessor accessor (table, "TSMExample");
    ROArrayColumn<float> freq (table, "Freq");
    ROArrayColumn<float> pol (table, "Pol");
    ROArrayColumn<float> data (table, "Data");
    ROArrayColumn<float> weight (table, "Weight");
    ROScalarColumn<float> time (table, "Time");
    Vector<float> freqValues(20);
    Vector<float> polValues(16);
    indgen (freqValues.ac(), float(200));
    indgen (polValues.ac(), float(300));
    float timeValue;
    timeValue = 34;
    Matrix<float> array(IPosition(2,16,20));
    Matrix<float> result(IPosition(2,16,20));
    uInt i;
    indgen (array.ac());
    for (i=0; i<table.nrow(); i++) {
	data.get (i, result);
	if (! allEQ (array.ac(), result.ac())) {
	    cout << "mismatch in data row " << i << endl;
	}
	weight.get (i, result);
	if (! allEQ (array.ac() + float(100), result.ac())) {
	    cout << "mismatch in weight row " << i << endl;
	}
	if (! allEQ (freq(i), freqValues.ac())) {
	    cout << "mismatch in freq row " << i << endl;
	}
	if (! allEQ (pol(i), polValues.ac())) {
	    cout << "mismatch in pol row " << i << endl;
	}
	if (time(i) != timeValue) {
	    cout << "mismatch in time row " << i << endl;
	}
	array.ac() += float(200);
	timeValue += 5;
    }
    accessor.showCacheStatistics (cout);
    accessor.clearCaches();
    cout << "get's have been done" << endl;

    // Get the entire column.
    // (In a separate scope to delete the big array at the end).
    {
	Array<float> result;
	data.getColumn (result);
	if (result.shape() != IPosition (3,16,20,51)) {
	    cout << "shape of getColumn result is incorrect" << endl;
	}else{
	    indgen (array.ac());
	    uInt i=0;
	    ArrayIterator<float> iter (result, 2);
	    while (! iter.pastEnd()) {
		if (! allEQ (iter.array(), array.ac())) {
		    cout << "mismatch in getColumn data row " << i << endl;
		}
		array.ac() += float(200);
		i++;
		iter.next();
	    }
	}
	accessor.showCacheStatistics (cout);
	accessor.clearCaches();
	cout << "getColumn has been done" << endl;
    }

    // Get slices in the entire column.
    {
	uInt i, j;
	Cube<float> array(1,1,51);
	for (i=0; i<51; i++) {
	    array(0,0,i) = 200*i;
	}
	Array<float> result;
	for (j=0; j<20; j++) {
	    for (i=0; i<16; i++) {
		data.getColumn (Slicer (IPosition(2,i,j)), result);
		if (! allEQ (result, array.ac())) {
		    cout << "mismatch in getColumnSlice " << i << "," << j
			 << endl;
		}
		array.ac() += float(1);
	    }
	}
	accessor.showCacheStatistics (cout);
	accessor.clearCaches();
	cout << "getColumnSlice's have been done" << endl;
    }

    // Get strided slices in the entire column.
    {
	uInt i, j;
	Cube<float> array(2,2,51);
	for (i=0; i<51; i++) {
	    array(0,0,i) = 200*i;
	    array(1,0,i) = 200*i + 16/2;
	    array(0,1,i) = 200*i + 16*20/2;
	    array(1,1,i) = 200*i + 16*20/2 + 16/2;
	}
	Array<float> result;
	for (j=0; j<20/2; j++) {
	    for (i=0; i<16/2; i++) {
		data.getColumn (Slicer (IPosition(2,i,j),
					IPosition(2,2,2),
					IPosition(2,16/2,20/2)),
				result);
		if (! allEQ (result, array.ac())) {
		    cout << "mismatch in getColumnSlice " << i << "," << j
			 << endl;
		}
		array.ac() += float(1);
	    }
	    array.ac() += float(8);
	}
	accessor.showCacheStatistics (cout);
	accessor.clearCaches();
	cout << "strided getColumnSlice's have been done" << endl;
    }

    // Get slices from each cell.
    {
	uInt i, j, k;
	Matrix<float> array(2,4);
	Array<float> result;
	for (k=0; k<table.nrow(); k++) {
	    for (j=0; j<4; j++) {
		for (i=0; i<2; i++) {
		    array(i,j) = float(i + 16*j + 200*k);
		}
	    }
	    for (j=0; j<20/4; j++) {
		for (i=0; i<16/2; i++) {
		    data.getSlice (k, 
				   Slicer (IPosition(2,i*2,j*4),
					   IPosition(2,2,4)),
				   result);
		    if (! allEQ (result, array.ac())) {
			cout << "mismatch in getSlice " << i << "," << j
			     << "," << k << endl;
		    }
		    array.ac() += float(2);
		}
		array.ac() += float((4-1) * 16);
	    }
	}
	accessor.showCacheStatistics (cout);
	accessor.clearCaches();
	cout << "getSlice's have been done" << endl;
    }

    // Get a strided slice from each cell.
    {
	uInt i, j, k;
	Matrix<float> array(8,5);
	Array<float> result;
	for (k=0; k<table.nrow(); k++) {
	    for (j=0; j<5; j++) {
		for (i=0; i<8; i++) {
		    array(i,j) = float(i*16/8 + 16*20/5*j + 200*k);
		}
	    }
	    for (j=0; j<20/5; j++) {
		for (i=0; i<16/8; i++) {
		    data.getSlice (k, 
				   Slicer (IPosition(2,i,j),
					   IPosition(2,8,5),
					   IPosition(2,16/8,20/5)),
				   result);
		    if (! allEQ (result, array.ac())) {
			cout << "mismatch in strided getSlice " << i << ","
			     << j << "," << k << endl;
		    }
		    array.ac() += float(1);
		}
		array.ac() += float(16 - 16/8);
	    }
	}
	accessor.showCacheStatistics (cout);
	accessor.clearCaches();
	cout << "getSlice's with strides have been done" << endl;
    }
}
