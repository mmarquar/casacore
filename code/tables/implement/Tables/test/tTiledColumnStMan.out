>>> TSMCube cache statistics:
cubeShape: [16, 20, 51]
tileShape: [5, 6, 1]
maxCacheSz:1048576
cacheSize: 16 (*240)
#accesses: 3264
#reads:    816
#inits:    816
#writes:   816
hit-rate:  50%
<<<
get's have been done
getColumn has been done
getColumnSlice's have been done
strided getColumnSlice's have been done
getSlice's have been done
getSlice's with strides have been done
>>> TSMCube cache statistics:
cubeShape: [16, 20, 51]
tileShape: [5, 6, 1]
maxCacheSz:1048576
cacheSize: 816 (*240)
#accesses: 43656
#reads:    1616
#inits:    0
#writes:   0
hit-rate:  96.2983%
<<<
